### ISSUE TITLE
Clean big files out of the git history

### ISSUE THREAD

**Issue by kenden** [rdoeffinger/Dictionary#111]
# Issue:
cloning the git repository takes a long time with a not-so-good-connection (126MB to download)

# Reason:
There are a lot of big files in the git history that are not part of the code anymore. 

The code itself is only 13MB
```bash
$ du -hs --exclude .git
13M
```
but the git history is much bigger
```bash
$ du -hs .git
113M

```
Using this command (from https://stackoverflow.com/a/42544963)
to find files in history bigger than 1M and **not** present in the HEAD of the git repo anymore:
```
$ git rev-list --objects --all \
| git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' \
| sed -n 's/^blob //p' \
| grep -vF --file=<(git ls-tree -r HEAD | awk '{print $3}') \
| awk '$2 >= 2^20' \
| sort --numeric-sort --key=2 \
| cut -c 1-12,41- \
| $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest
f31de9a2a9a2  1.0MiB jars/icu4j-4_2_1-src/src/com/ibm/icu/dev/test/perf/data/collation/TestNames_SerbianSH.txt
607033e2db7f  1.1MiB jars/icu4j-4_2_1-src/src/com/ibm/icu/dev/data/unicode/UnicodeData.txt
cad86e4aea76  1.2MiB jars/icu4j-4_4_2-src/main/tests/core/src/com/ibm/icu/dev/data/unicode/UnicodeData.txt
8d7222b13789  1.3MiB jars/icu4j-4_8_1_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/UnicodeData.txt
9fffa71a1e92  1.3MiB jars/icu4j-52_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/UnicodeData.txt
f89f7b00a6f5  1.3MiB jars/icu4j-52_1/main/shared/data/Transliterator_Han_Latin_Definition.txt
17e8e1a152bb  1.4MiB jars/icu4j-4_4_2-src/main/shared/data/Transliterator_Han_Latin_Definition.txt
f082442b2f4b  1.5MiB images/flags/Flag_of_Serbia.svg
2072cdaf4855  1.5MiB assets/FreeSans.ttf.jpg
b322c280fc88  1.6MiB images/flags.xcf
b8906f505ea2  1.6MiB assets/FreeSerif.ttf.jpg
4863e253f044  1.6MiB jars/icu4j-4_8_1_1/main/tests/collate/src/com/ibm/icu/dev/data/CollationTest_NON_IGNORABLE_SHORT.txt
28a6a81d8acb  1.7MiB jars/icu4j-52_1/perf-tests/data/collation/TestNames_SerbianSR.txt
f2a7ec0ffb7a  1.7MiB jars/icu4j-52_1/main/tests/collate/src/com/ibm/icu/dev/data/CollationTest_NON_IGNORABLE_SHORT.txt
4ea10150d356  1.7MiB jars/icu4j-4_4_2-src/main/tests/collate/src/com/ibm/icu/dev/data/CollationTest_SHIFTED_SHORT.txt
006d05cc2d21  1.7MiB jars/icu4j-4_4_2-src/main/tests/collate/src/com/ibm/icu/dev/data/CollationTest_NON_IGNORABLE_SHORT.txt
a34711dc66c3  1.7MiB jars/icu4j-4_2_1-src/src/com/ibm/icu/dev/test/perf/data/collation/TestNames_SerbianSR.txt
2c7ecf116a30  1.7MiB jars/icu4j-4_8_1_1/main/tests/collate/src/com/ibm/icu/dev/data/CollationTest_SHIFTED_SHORT.txt
c155d132c3ab  1.8MiB jars/icu4j-52_1/main/tests/collate/src/com/ibm/icu/dev/data/CollationTest_SHIFTED_SHORT.txt
c303b5a2871f  2.2MiB jars/icu4j-4_8_1_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/confusables.txt
cbd88c57aaf2  2.2MiB jars/icu4j-52_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/confusables.txt
38c00263dc8a  2.5MiB jars/icu4j-52_1/main/shared/data/Transliterator_Han_Latin_EDICT.txt
49d75718339b  2.6MiB jars/icu4j-4_4_2-src/main/shared/data/Transliterator_Han_Latin_EDICT.txt
889c594f6674  3.2MiB assets/FreeSerif.ttf.jpg
675d15bd4ab7  3.2MiB jars/icu4j-4_8_1_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/BidiTest.txt
56d35c88fde4  3.6MiB jars/icu4j-4_4_2-src/icu4j.jar
6a6153e26507  3.6MiB jars/icu4j-module.jar
e6469ef0730f  5.1MiB Dictionary_signed.apk
9f9a30f9411f  5.1MiB Dictionary_signed.apk
435dfa092d0c  5.1MiB Dictionary_signed.apk
80e813638d74  5.5MiB jars/icu4j-module.jar
7991959d6872  6.0MiB Dictionary_signed.apk
7a4ec7fbfe55  6.3MiB Dictionary_signed.apk
9207236938d6  6.4MiB Dictionary_signed.apk
82f68eb9b8ed  6.4MiB Dictionary_signed.apk
73ca581c3481  6.6MiB jars/icu4j-52_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/BidiCharacterTest.txt
b23391382a95  6.7MiB jars/icu4j-4_4_2-src/main/shared/data/icudata.jar
5e304dd53478  6.9MiB jars/icu4j-4_2_1-src/src/com/ibm/icu/impl/data/icudata.jar
02c2c9d177af  7.6MiB jars/icu4j-52_1/main/tests/core/src/com/ibm/icu/dev/data/unicode/BidiTest.txt
dfbafe092b5e  7.6MiB jars/icu4j-4_8_1_1/main/shared/data/icudata.jar
28fafed8713a  9.1MiB Dictionary_signed.apk
58cd7d318da9  9.1MiB Dictionary_signed.apk
fd7ad592b4ca  9.1MiB Dictionary_signed.apk
6b3c9ea13ad6   10MiB jars/icu4j-52_1/main/shared/data/icudata.jar

```

It seems that for a lot of those files (all?), there is no value in keeping them in the git history.

# Suggestion:
Remove those files from the git history. 
Note this is a bit slow (but only once!) and this is a **_DESTRUCTIVE_** operation: _**do a backup**_ first.
From https://stackoverflow.com/questions/2100907/how-to-remove-delete-a-large-file-from-commit-history-in-git-repository
We can get commands to clean up those files.
e.g., for the last file:
```bash
$ git filter-branch --force --index-filter 'git rm --cached -r --ignore-unmatch jars/icu4j-52_1/main/shared/data/icudata.jar' --prune-empty --tag-name-filter cat -- --all
```
(repeat for all files)
finish with a `git push --force`

**Comment by rdoeffinger** [comment]
Sorry, but re-writing git history will break every single clone or fork out there, and fixing that would be a major pain for everyone.
So as long as the repo is not completely and utterly broken, doing this is simply a no-go.
It is possible to clone without history (advisable on slow connections).
It should also be possible to specify to git to switch to an alternative tree at some point  (junction points or something like that?) so you could have history from such a cleaned-up branch without needing to download the full main branch. But I expect that this will be quite some work and tricky to do. But if it is useful I'm happy to add a SEPARATE branch (that will not be updated) with the cleaned up history, allowing someone to get master for recent history and that branch for old history.

**Comment by kenden** [comment]
Thanks for the answer. It makes sense, not need to do a separate branch.
I'm closing this, I don't think it has to stay open.
